pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,753 @@
+				PortUUID: interfaceConfig.UplinkPort.PortUUID,
+				OFPort:   interfaceConfig.UplinkPort.OFPort,
+			},
+			HostIfaceIndex: interfaceConfig.EntityInterfaceConfig.HostIfaceIndex,
+		},
+		IPs: ifIPs,
+	}
+	return iface, nil
+}
+
+func (c *ExternalNodeController) removeOVSPortsAndFlows(interfaceConfig *interfacestore.InterfaceConfig) error {
+	portUUID := interfaceConfig.PortUUID
+	portName := interfaceConfig.InterfaceName
+	if err := c.ofClient.UninstallVMUplinkFlows(portName); err != nil {
+		return err
+	}
+	port, err := c.ovsBridgeClient.GetPortData(portUUID, portName)
+	if err != nil {
+		if strings.Contains(err.Error(), "not found") {
+			klog.ErrorS(err, "OVS port is not found", "port", portName)
+			return nil
+		}
+		return err
+	}
+	hostIfName := port.ExternalIDs[ovsExternalIDHostIFName]
+	uplinkIfName := port.ExternalIDs[ovsExternalIDUplinkName]
+	uplinkPortID := port.ExternalIDs[ovsExternalIDUplinkPort]
+	adapterConfig, er := util.GetUplinkConfig(hostIfName)
+	if er != nil {
+		klog.ErrorS(err, "Failed to get the configuration on the host interface", "hostInterface", hostIfName)
+		return err
+	}
+	if err := c.ovsBridgeClient.DeletePort(port.UUID); err != nil {
+		klog.ErrorS(err, "Failed to delete port on OVS", "port", hostIfName)
+		return err
+	}
+	if err := c.ovsBridgeClient.DeletePort(uplinkPortID); err != nil {
+		return err
+	}
+	defer func() {
+		// Delete host interface from OVS datapath if it is existing.
+		// This is to resolve an issue that OVS fails to remove the interface from datapath. It might happen because the interface
+		// is busy when OVS tries to remove it with the OVSDB interface deletion event.
+		if err := c.ovsctlClient.DeleteDPInterface(hostIfName); err != nil {
+			klog.ErrorS(err, "Failed delete host interface from OVS datapath", "interface", hostIfName)
+		}
+	}()
+
+	// Wait until the host interface created by OVS is removed.
+	pollErr := wait.PollImmediate(50*time.Millisecond, 2*time.Second, func() (bool, error) {
+		return !util.HostInterfaceExists(hostIfName), nil
+	})
+	if pollErr != nil {
+		klog.ErrorS(pollErr, "Failed to wait for host interface deletion in 2s", "interface", hostIfName)
+	}
+	// Recover the uplink interface's name.
+	if err := util.RenameInterface(uplinkIfName, hostIfName); err != nil {
+		klog.ErrorS(err, "Failed to recover uplink name to the host interface name", "uplink", uplinkIfName, "hostInterface", hostIfName)
+		return err
+	}
+	// Move the IP configurations back to the host interface.
+	if err := util.MoveIFConfigurations(adapterConfig.IPs, adapterConfig.Routes, adapterConfig.MAC, adapterConfig.MTU, uplinkIfName, hostIfName); err != nil {
+		klog.ErrorS(err, "Failed to move back configuration to the host interface", "hostInterface", hostIfName)
+		return err
+	}
+	return nil
+}
+
+func getHostInterfaceName(iface v1alpha1.NetworkInterface) (string, error) {
+	var ipFilter *ip.DualStackIPs
+	epIP := net.ParseIP(iface.IPs[0])
```

Comment:
Is it necessary to have this function in externalnode package? Seems util package is the appropriate place

Reply:
Thanks for pointing out. I have changed to process all the IPs. I did not move this to util because it is not used by other packages, but only this file.